File: sdk/runanywhere-react-native/packages/core/src/types/VoiceAgentTypes.ts
/**
 * VoiceAgentTypes.ts
 *
 * Type definitions for Voice Agent functionality.
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/VoiceAgent/
 */

// v3.1: proto imports removed — legacy mapper helpers that used them
// (voiceSessionEventFromProto / voiceSessionEventKindFromProto) were
// deleted. Consumers import VoiceEvent directly from
// '@runanywhere/proto-ts/voice_events' when they need it.

/**
 * Component load state
 */
export type ComponentLoadState = 'notLoaded' | 'loading' | 'loaded' | 'failed';

/**
 * Individual component state
 */
export interface ComponentState {
  state: ComponentLoadState;
  modelId?: string;
  voiceId?: string;
}

/**
 * Voice agent component states
 */
export interface VoiceAgentComponentStates {
  stt: ComponentState;
  llm: ComponentState;
  tts: ComponentState;
  isFullyReady: boolean;
}

/**
 * Voice agent configuration
 */
export interface VoiceAgentConfig {
  /** STT model ID */
  sttModelId?: string;

  /** LLM model ID */
  llmModelId?: string;

  /** TTS voice ID */
  ttsVoiceId?: string;

  /** VAD sample rate (default: 16000) */
  vadSampleRate?: number;

  /** VAD frame length (default: 512) */
  vadFrameLength?: number;

  /** VAD energy threshold (default: 0.1) */
  vadEnergyThreshold?: number;

  /** Language code (e.g., 'en') */
  language?: string;

  /** System prompt for LLM */
  systemPrompt?: string;
}

/**
 * Voice turn result
 */
export interface VoiceTurnResult {
  /** Whether speech was detected */
  speechDetected: boolean;

  /** Transcribed text from audio */
  transcription: string;

  /** Generated response text */
  response: string;

  /** Base64-encoded synthesized audio */
  synthesizedAudio?: string;

  /** Audio sample rate */
  sampleRate: number;
}

// v3.1: VoiceSessionEvent / VoiceSessionEventType interface +
// voiceSessionEventFromProto + voiceSessionEventKindFromProto mappers +
// VoiceSessionCallback DELETED. Use VoiceEvent (ts-proto) via
// VoiceAgentStreamAdapter.stream() directly.

/**
 * Voice agent metrics
 */
export interface VoiceAgentMetrics {
  /** Time for STT processing (ms) */
  sttLatencyMs: number;

  /** Time for LLM generation (ms) */
  llmLatencyMs: number;

  /** Time for TTS synthesis (ms) */
  ttsLatencyMs: number;

  /** Total turn latency (ms) */
  totalLatencyMs: number;

  /** Number of tokens generated */
  tokensGenerated: number;

  /** Audio duration (seconds) */
  audioDurationSeconds: number;
}

// v3.1: VoiceSessionConfig / VoiceSessionEventKind / VoiceSessionErrorType
// DELETED. Use VoiceAgentConfig + VoiceEvent (ts-proto) directly via
// VoiceAgentStreamAdapter.

File: sdk/runanywhere-flutter/packages/runanywhere/lib/public/types/voice_agent_types.dart
/// Voice Agent Types
///
/// Types for voice agent operations.
/// Matches Swift VoiceAgentTypes.swift from Public/Extensions/VoiceAgent/
library voice_agent_types;

// MARK: - Component Load State

/// State of a voice agent component
sealed class ComponentLoadState {
  const ComponentLoadState();

  /// Component is not loaded
  const factory ComponentLoadState.notLoaded() = ComponentLoadStateNotLoaded;

  /// Component is currently loading.
  const factory ComponentLoadState.loading() = ComponentLoadStateLoading;

  /// Component is loaded with the given model ID
  const factory ComponentLoadState.loaded({required String modelId}) =
      ComponentLoadStateLoaded;

  /// Component failed to load with an error message.
  const factory ComponentLoadState.error(String message) =
      ComponentLoadStateError;
}

/// Component not loaded state
class ComponentLoadStateNotLoaded extends ComponentLoadState {
  const ComponentLoadStateNotLoaded();
}

/// Component is currently loading.
class ComponentLoadStateLoading extends ComponentLoadState {
  const ComponentLoadStateLoading();
}

/// Component loaded state
class ComponentLoadStateLoaded extends ComponentLoadState {
  /// ID of the loaded model
  final String modelId;

  const ComponentLoadStateLoaded({required this.modelId});
}

/// Component encountered a load error.
class ComponentLoadStateError extends ComponentLoadState {
  /// Failure reason / message.
  final String message;

  const ComponentLoadStateError(this.message);
}

// MARK: - Voice Agent Component States

/// States of all voice agent components (STT, LLM, TTS)
///
/// Matches Swift VoiceAgentComponentStates from VoiceAgentTypes.swift
class VoiceAgentComponentStates {
  /// Speech-to-Text component state
  final ComponentLoadState stt;

  /// Large Language Model component state
  final ComponentLoadState llm;

  /// Text-to-Speech component state
  final ComponentLoadState tts;

  const VoiceAgentComponentStates({
    this.stt = const ComponentLoadState.notLoaded(),
    this.llm = const ComponentLoadState.notLoaded(),
    this.tts = const ComponentLoadState.notLoaded(),
  });

  /// Check if all components are loaded
  bool get isFullyReady =>
      stt is ComponentLoadStateLoaded &&
      llm is ComponentLoadStateLoaded &&
      tts is ComponentLoadStateLoaded;

  /// Check if any component is loaded
  bool get hasAnyLoaded =>
      stt is ComponentLoadStateLoaded ||
      llm is ComponentLoadStateLoaded ||
      tts is ComponentLoadStateLoaded;

  /// True if any component is currently loading.
  bool get isAnyLoading =>
      stt is ComponentLoadStateLoading ||
      llm is ComponentLoadStateLoading ||
      tts is ComponentLoadStateLoading;

  /// Names of components that are not yet loaded.
  List<String> get missingComponents {
    final missing = <String>[];
    if (stt is! ComponentLoadStateLoaded) missing.add('stt');
    if (llm is! ComponentLoadStateLoaded) missing.add('llm');
    if (tts is! ComponentLoadStateLoaded) missing.add('tts');
    return missing;
  }

  @override
  String toString() {
    String stateToString(ComponentLoadState state) {
      if (state is ComponentLoadStateLoaded) {
        return 'loaded(${state.modelId})';
      }
      if (state is ComponentLoadStateLoading) return 'loading';
      if (state is ComponentLoadStateError) return 'error(${state.message})';
      return 'notLoaded';
    }

    return 'VoiceAgentComponentStates('
        'stt: ${stateToString(stt)}, '
        'llm: ${stateToString(llm)}, '
        'tts: ${stateToString(tts)})';
  }
}
